### DISCUSSION TITLE
Is there any way to make the process faster?

### DISCUSSION THREAD

**Discussion by kimkihoon0515** [fastapi/fastapi#8516] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, File, UploadFile
from prediction import predict
import uvicorn
import numpy as np
from typing import List
import cv2
from image_crop import crop
from grad_cam import GradCam,show_cam_on_image,preprocess_image
from efficientnet_pytorch import EfficientNet
from fastapi import FastAPI
import torch
import time

app = FastAPI()

model_path = [] # 5 model paths in here.
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model0 = EfficientNet.from_pretrained('efficientnet-b0',num_classes=5).to(device)
model0.load_state_dict(torch.load(model_path[0],map_location=device))
model1 = EfficientNet.from_pretrained('efficientnet-b0',num_classes=5).to(device)
model1.load_state_dict(torch.load(model_path[1],map_location=device))
model2 = EfficientNet.from_pretrained('efficientnet-b0',num_classes=5).to(device)
model2.load_state_dict(torch.load(model_path[2],map_location=device))
model3 = EfficientNet.from_pretrained('efficientnet-b0',num_classes=5).to(device)
model3.load_state_dict(torch.load(model_path[3],map_location=device))
model4 = EfficientNet.from_pretrained('efficientnet-b0',num_classes=5).to(device)
model4.load_state_dict(torch.load(model_path[4],map_location=device))


@app.post('/gradcam')
async def get_gradcam(files: List[UploadFile] = File(...)):
    global model0,model1,model2,model3,model4,model_path,device
    image_bytes = await files[0].read() 
    encoded_img = np.frombuffer(image_bytes,dtype=np.uint8)
    new_img = cv2.imdecode(encoded_img,cv2.IMREAD_COLOR)

    crop_img = crop(new_img)
    cam_list = []
    label_list = []
    start= time.time()
    for i in range(5):
        if i == 0:
            model = model0
        elif i == 1:
            model = model1
        elif i == 2:
            model = model2
        elif i == 3:
            model = model3
        else:
            model = model4
        #model.load_state_dict(torch.load(model_path[i],map_location=device))
        label = predict(model=model,img=crop_img) # getting labels
        label_list.append(label)
        model = model.to('cpu')  
        rgb_img=cv2.cvtColor(crop_img, cv2.COLOR_RGB2BGR) 
        rgb_img = np.float32(rgb_img) / 255

        cam_img = preprocess_image(rgb_img)
        # target layer
        target_layer_names = [str(len(model._blocks) - 1)]
        img_size = (rgb_img.shape[1], rgb_img.shape[0])
        grad_cam = GradCam(model=model, blob_name = '_blocks', target_layer_names=target_layer_names, use_cuda=True,img_size=img_size)
        target_index = label
        mask_dic = grad_cam(cam_img, target_index)
        for k,v in mask_dic.items():
            gray_mask=v
            name=k
        cam_image = show_cam_on_image(rgb_img, 1-gray_mask, name) # getting grad-cam result
        cam_list.append(cam_image.tolist())
    end = time.time()
    print(end-start)
    return {'cam':cam_list,'label':label_list}
```

### Description

I'm trying to post 1 image to FastAPI server and get 5 labels and  grad-cam results. But his process takes too long to get results. Is there any way to make it faster? 

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

Python 3.10.4

**Comment by JarroVGIT** [comment]
This is not related to FastAPI but to general code quality. It isn't FastAPI that is slow, it is the business logic you added yourself. 

General tips:

- Add timers to see what in your code actually takes a long time;
- Replace the following:
```python
    for i in range(5):
        if i == 0:
            model = model0
        elif i == 1:
            model = model1
        elif i == 2:
            model = model2
        elif i == 3:
            model = model3
        else:
            model = model4

        label = predict(model=model,img=crop_img)
```
with something more pythonic like this:
```python
models = [model0, model1, model2, model3, model4]
for model in models:
    label = predict(model=model,img=crop_img)
```
- I am not sure what you are trying to do here, but this would only give you the last read k/v of the dictionary to the variables `graymask` and `name`:
```python
for k,v in mask_dic.items():
            gray_mask=v
            name=k
```
- Finally, if you have measured what part of your code takes too long and you have determined for yourself what an acceptable timeframe is for it to be executed in, ask very specific questions on Stackoverflow. In general, questions as 'how do I make this faster' are closed immediately there. 

Good luck!

**Comment by kalmastenitin** [comment]
why you are using list of upload files? You are using only single image file. This is unnessery. Also its very good suggestion by @JarroVGIT. Your model loading and image analysis is time consuming process here.

**Comment by kimkihoon0515** [comment]
@kalmastenitin Trying to use files: UploadFile = File(...) but got an error. Currently working on it. Thx for the tip btw :D

**Comment by sm-Fifteen** [comment]
It's also probably ill-advised to do any significant amount of processing in your FastAPI route, especially if the function is marked as `async`, because it will block your main thread for the runtime of the function. You might want to move `image_bytes`, `encoded_img` and `new_img` into a separate function and then call it as a dependency in your route, to allow you to make the function non-async and make it run in its own thread:

```py
async def gradcam_image_data(files: List[UploadFile] = File(...)):
    image_bytes = await files[0].read() 
    encoded_img = np.frombuffer(image_bytes,dtype=np.uint8)
    new_img = cv2.imdecode(encoded_img,cv2.IMREAD_COLOR)
    return new_img

# not async anymore
@app.post('/gradcam')
def get_gradcam(new_img = Depends(gradcam_image_data)):
    global model0,model1,model2,model3,model4,model_path,device
    
    crop_img = crop(new_img)
    ...
```

I'd imagine `predict` and possibly `preprocess_image` or `grad_cam` must take a while to run, so moving them to their own thread might help (assuming these don't also hold the GIL, which would still prevent other threads from running during the processing phase).

**Comment by drcode-rune** [comment]
Why are you doing this?

```Python
for k,v in mask_dic.items():
        gray_mask=v
        name=k
cam_image = show_cam_on_image(rgb_img, 1-gray_mask, name) # getting grad-cam result
cam_list.append(cam_image.tolist())
```

It will always use the last key-value pair so in `cam_image = show_cam_on_image(rgb_img, 1-gray_mask, name)` you will be operating with this last pair not with each pair, if you want to use each key-value pair for `show_cam_on_image` you should do something like this instead

```Python
for k,v in mask_dic.items():
        cam_image = show_cam_on_image(rgb_img, 1-v, k) # getting grad-cam result. v = gray_mask, k = name
        cam_list.append(cam_image.tolist())
```

**Comment by prabal01pathak** [comment]
I would suggest going for multiprocessing using some queue-based architecture
like: zmq or simply you can use multiprocessing module.
